### PR TITLE
Avoid panicking in Usage Plugin and ensure valid supergraph SDL

### PR DESCRIPTION
**Pull request by kamilkisiela**

### CODE REVIEW

File: packages/libraries/router/src/agent.rs [hunk 1]
```diff
@@ -112,6 +122,20 @@ fn non_empty_string(value: Option<String>) -> Option<String> {
     }
 }
 
+#[derive(Error, Debug)]
+pub enum AgentError {
+    #[error("unable to acquire lock for {0} in {1}, reason {2}")]
+    Lock(String, String, String),
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/agent.rs:128]:
```suggestion
    #[error("unable to acquire lock: {0}")]
    Lock(String),
```

I don't think stating the struct and the method name is necessary, bc the error would include the error line position anyways.

ex:
```sh
2023-10-19T17:40:51.450438Z  ERROR panicked at 'oh no, what happened?', packages/libraries/router/src/agent.rs:201:17
```

File: packages/libraries/router/src/agent.rs [hunk 2]
```diff
@@ -173,13 +197,25 @@ impl UsageAgent {
...
                 .processor
                 .lock()
-                .expect("Unable to acquire the OperationProcessor in produce_report")
+                .map_err(|e| {
+                    AgentError::Lock(
+                        "OperationProcessor".to_string(),
+                        "produce_report".to_string(),
+                        e.to_string(),
+                    )
+                })?
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/agent.rs:206]:
```suggestion
                .map_err(|e| {
                    AgentError::Lock(
                        e.to_string(),
                    )
                })?
```

File: packages/libraries/router/src/agent.rs [hunk 3]
```diff
@@ -173,13 +197,25 @@ impl UsageAgent {
...
                         .state
                         .lock()
-                        .expect("Unable to acquire State in produce_report")
+                        .map_err(|e| {
+                            AgentError::Lock(
+                                "State".to_string(),
+                                "produce_report".to_string(),
+                                e.to_string(),
+                            )
+                        })?
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/agent.rs:218]:
```suggestion
                        .map_err(|e| {
                            AgentError::Lock(
                                e.to_string(),
                            )
                        })?
```

File: packages/libraries/router/src/agent.rs [hunk 4]
```diff
@@ -233,24 +269,28 @@ impl UsageAgent {
...
 
-    pub fn add_report(&self, execution_report: ExecutionReport) {
+    pub fn add_report(&self, execution_report: ExecutionReport) -> Result<(), AgentError> {
         let size = self
             .state
             .lock()
-            .expect("Unable to acquire State in add_report")
+            .map_err(|e| {
+                AgentError::Lock("State".to_string(), "add_report".to_string(), e.to_string())
+            })?
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/agent.rs:281]:
```suggestion
            .map_err(|e| {
                AgentError::Lock(e.to_string())
            })?
```

File: packages/libraries/router/src/usage.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/usage.rs:1]:
```diff
impl Default for Config {
    fn default() -> Self {
        Self {
            enabled: Some(true),
            sample_rate: Some(1.0),
            exclude: None,
-           client_name_header: None,
-           client_version_header: None,
+           client_name_header: Some(String::from("graphql-client-name")),
+           client_version_header: Some(String::from("graphql-client-version")),
            accept_invalid_certs: Some(false),
            buffer_size: Some(1000),
            connect_timeout: Some(5),
            request_timeout: Some(15),
        }
    }
}
```

File: packages/libraries/router/src/usage.rs [hunk 5]
```diff
@@ -177,19 +172,48 @@ impl Plugin for UsagePlugin {
...
         }
 
-
         Ok(UsagePlugin {
-            config: init.config,
+            config: OperationConfig {
+                sample_rate: user_config
+                    .sample_rate
+                    .or(default_config.sample_rate)
+                    .expect("sample_rate has default value"),
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/usage.rs:206]:
```suggestion
                    .expect("sample_rate has no default value"),
```

File: packages/libraries/router/src/usage.rs [hunk 6]
```diff
@@ -177,19 +172,48 @@ impl Plugin for UsagePlugin {
...
+            .or(default_config.accept_invalid_certs)
+            .expect("accept_invalid_certs has default value");
+        let connect_timeout = user_config
+            .connect_timeout
+            .or(default_config.connect_timeout)
+            .expect("connect_timeout has default value");
+        let request_timeout = user_config
+            .request_timeout
+            .or(default_config.request_timeout)
+            .expect("request_timeout has default value");
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/usage.rs:195]:
```suggestion
            .expect("enabled has default value");
        let buffer_size = user_config
            .buffer_size
            .or(default_config.buffer_size)
            .expect("buffer_size has no default value");
        let accept_invalid_certs = user_config
            .accept_invalid_certs
            .or(default_config.accept_invalid_certs)
            .expect("accept_invalid_certs has no default value");
        let connect_timeout = user_config
            .connect_timeout
            .or(default_config.connect_timeout)
            .expect("connect_timeout has no default value");
        let request_timeout = user_config
            .request_timeout
            .or(default_config.request_timeout)
            .expect("request_timeout has no default value");
```

File: packages/libraries/router/src/usage.rs [hunk 7]
```diff
@@ -177,19 +172,48 @@ impl Plugin for UsagePlugin {
...
+            config: OperationConfig {
+                sample_rate: user_config
+                    .sample_rate
+                    .or(default_config.sample_rate)
+                    .expect("sample_rate has default value"),
+                exclude: user_config.exclude.or(default_config.exclude),
+                client_name_header: user_config
+                    .client_name_header
+                    .or(default_config.client_name_header)
+                    .expect("client_name_header has default value"),
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/usage.rs:211]:
```suggestion
                    .expect("client_name_header has no default value"),
```

File: packages/libraries/router/src/usage.rs [hunk 8]
```diff
@@ -177,19 +172,48 @@ impl Plugin for UsagePlugin {
...
+                client_version_header: user_config
+                    .client_version_header
+                    .or(default_config.client_version_header)
+                    .expect("client_version_header has default value"),
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/usage.rs:215]:
```suggestion
                    .expect("client_version_header has no default value"),
```

File: packages/libraries/router/src/usage.rs [hunk 9]
```diff
@@ -211,100 +235,126 @@ impl Plugin for UsagePlugin {
...
 
+fn try_add_report(agent: Arc<Mutex<UsageAgent>>, execution_report: ExecutionReport) {
+    agent
+        .lock()
+        .map_err(|e| {
+            AgentError::Lock(
+                "Agent".to_string(),
+                "supergraph_service".to_string(),
+                e.to_string(),
+            )
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/usage.rs:350]:
```suggestion
AgentError::Lock(e.to_string())
```

**Review by dotansimha — APPROVED** [review]

**Review by YassinEldeeb — APPROVED** [review]

**Pull request merged**